### PR TITLE
feat: DataGrid - add option to allow columns overflowing the container

### LINKED ISSUE #30379: [Bug]: DataGrid resizing columns is not possible when overflowX is enabled
### Library

React Components / v9 (@fluentui/react-components)

### System Info

```shell
System:
    OS: Windows 10 10.0.19045
    CPU: (12) x64 Intel(R) Core(TM) i7-8700K CPU @ 3.70GHz
    Memory: 13.71 GB / 31.78 GB
  Browsers:
    Edge: Spartan (44.19041.3570.0), Chromium (120.0.2210.144)
    Internet Explorer: 11.0.19041.3570
```

### Are you reporting Accessibility issue?

None

### Reproduction

https://codesandbox.io/p/sandbox/romantic-bardeen-rr4h28?file=%2Fsrc%2FApp.tsx%3A16%2C20

## Actual Behavior
chrome-capture-2024-1-24

## Expected Behavior
DetailsList Component FluentUI v8:
chrome-capture-2024-1-24 (2)

I think this is important to be fixed because content inside cell is not readable.

### Logs

_No response_

### Requested priority

High

### Products/sites affected

_No response_

### Are you willing to submit a PR to fix?

no

### Validations

- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] The provided reproduction is a minimal reproducible example of the bug.

### PR DESCRIPTION
**Pull request by george-cz**
## Previous Behavior

Previously, when the container was narrower than the added column's minimum widths, user was not able to resize any of the columns, as that would break the "columns have to fit within the container" rule.

https://github.com/microsoft/fluentui/assets/2070479/6fe827c4-238f-4aef-9a03-d8e70f513410

## New Behavior

Allow the developer to pass `resizableColumnsOptions={{ allowContainerOverflow: true }}` prop to DataGrid, removing the constraint that columns always have to fit in the container.

As a side effect, this disables the dynamic adjacent column resizing even when the container is wide enough, since those two ideas collide.

https://github.com/microsoft/fluentui/assets/2070479/4d33f441-32e7-4073-a392-e2f8f07522d1

## Related Issue(s)

- Fixes #30379

### CODE REVIEW

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)
<p>No significant results to display.</p><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30717/perf-test-react-components/Avatar-mount_base.html">669</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30717/perf-test-react-components/Avatar-mount.html">620</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30717/perf-test-react-components/Button-mount_base.html">297</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30717/perf-test-react-components/Button-mount.html">296</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30717/perf-test-react-components/Field-mount_base.html">1166</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30717/perf-test-react-components/Field-mount.html">1123</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30717/perf-test-react-components/FluentProvider-mount_base.html">708</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30717/perf-test-react-components/FluentProvider-mount.html">685</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30717/perf-test-react-components/FluentProviderWithTheme-mount_base.html">79</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30717/perf-test-react-components/FluentProviderWithTheme-mount.html">75</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30717/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">57</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30717/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">65</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30717/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">74</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30717/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">73</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30717/perf-test-react-components/MakeStyles-mount_base.html">851</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30717/perf-test-react-components/MakeStyles-mount.html">868</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30717/perf-test-react-components/Persona-mount_base.html">1782</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30717/perf-test-react-components/Persona-mount.html">1727</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30717/perf-test-react-components/SpinButton-mount_base.html">1365</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/30717/perf-test-react-components/SpinButton-mount.html">1403</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/79d0865ec5c91b6e0b23ae0931a18a181b825b55'>79d0865ec5c91b6e0b23ae0931a18a181b825b55</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## 🕵 fluentuiv9 No visual regressions between this PR and main
<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

File: packages/react-components/react-table/src/hooks/useTableColumnResizeState.ts [hunk 1]
```diff
@@ -43,30 +43,32 @@ type ColumnResizeStateAction<T> =
...
+            : adjustColumnWidthsToFitContainer(state.columnWidthState, action.containerWidth),
         };
 
       case 'COLUMNS_UPDATED':
         const newS = columnDefinitionsToState(action.columns, state.columnWidthState, state.columnSizingOptions);
         return {
           ...state,
           columns: action.columns,
-          columnWidthState: adjustColumnWidthsToFitContainer(newS, state.containerWidth),
+          columnWidthState: allowOverflow ? newS : adjustColumnWidthsToFitContainer(newS, state.containerWidth),
```
Comment (ling1726) [inline packages/react-components/react-table/src/hooks/useTableColumnResizeState.ts:63]: Please add some tests for this in `useTableColumnResizeState.test.ts`

File: packages/react-components/react-table/stories/DataGrid/ResizableColumnsAllowContainerOverflow.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+    defaultWidth: 180,
+  },
+  author: {
+    defaultWidth: 180,
+    minWidth: 120,
+    idealWidth: 180,
+  },
+};
+
+export const ResizableColumns = () => {
```
Comment (ling1726) [inline packages/react-components/react-table/stories/DataGrid/ResizableColumnsAllowContainerOverflow.stories.tsx:177]: Let's add an equivalent example in Table

**Pull request merged**